aten/src/ATen/native/mps/OperationUtils.h

```diff
@@ -0,0 +1,196 @@
+//  Copyright © 2022 Apple Inc.
+
+#include <ATen/ATen.h>
+#include <ATen/Tensor.h>
+#include <ATen/Utils.h>
+#include <ATen/mps/MPSStream.h>
+#include <ATen/native/mps/TensorFactory.h>
+#include <c10/core/ScalarType.h>
+#include <torch/library.h>
+#include <unordered_map>
+
+#ifdef __OBJC__
+#include <MetalPerformanceShaders/MetalPerformanceShaders.h>
+#endif
+
+using namespace at::mps;
+
+namespace at {
+namespace native {
+namespace mps {
+
+struct TORCH_CUDA_CPP_API MPSGeneratorImpl : public c10::GeneratorImpl {
+  MPSGeneratorImpl(DeviceIndex device_index = -1);
+  ~MPSGeneratorImpl() = default;
+
+  void set_current_seed(uint64_t seed) override;
+  uint64_t current_seed() const override;
+  uint64_t seed() override;
+  void set_state(const c10::TensorImpl& new_state) override;
+  c10::intrusive_ptr<c10::TensorImpl> get_state() const override;
+  static DeviceType device_type();
+
+private:
+  MPSGeneratorImpl* clone_impl() const override;
+  uint64_t seed_ = default_rng_seed_val;
+};
+
+const Generator& getDefaultMPSGenerator();
+
+void runMPSGraph(
+    MPSStream* mpsStream,
+    MPSGraph* mpsGraph,
+    NSDictionary* feeds,
+    NSDictionary* results);
+
+MPSDataType getMPSDataType(ScalarType scalar_type);
+MPSDataType getMPSScalarType(ScalarType scalar_type);
+std::string getMPSTypeString(ScalarType scalar_type);
+std::string getMPSShapeString(MPSShape* shape);
+std::string getTensorsStringKey(const TensorList tensors);
+double getMPSScalarValue(const Tensor& t);
+std::string getArrayRefString(const IntArrayRef s);
+
+MPSShape* getMPSShape(const Tensor& t);
+MPSShape* getMPSShape(IntArrayRef sizes);
+MPSShape* getMPSShape(c10::MaybeOwned<Tensor> t);
+
+class Placeholder {
+ public:
+  Placeholder() : _placeholder(nullptr), _value(nullptr) {}
+  Placeholder(MPSGraphTensor* mpsGraphTensor) : _placeholder(mpsGraphTensor), _value(nullptr) {}
```

Comment:
The raw pointers here are making my eyes bleed a little. I don't know much about Objective C; you still get to use C++ features in it, is that right? Can we use smart pointers for ownership in the code here?

Reply:
In ObjC/C++ its common to have instance variables in ptrs. These could be wrapped in some smart pointer ownership. But if you noticed these are all wrapped in :
```
@autorelease {
}
```
Which does automatic cleanup. If we do want to retain ownership. We use:
```
  if (!_commandBuffer) {
    _commandBuffer =
        [MPSCommandBuffer commandBufferFromCommandQueue:_commandQueue].retain;
  }
```